crates/compiler/mono/src/tail_recursion.rs

```diff
@@ -323,3 +404,636 @@ fn insert_jumps<'a>(
+
+            accum
+        }
+        Stmt::Refcounting(_, next) => trmc_candidates_help(function_name, next, candidates),
+        Stmt::Expect { remainder, .. }
+        | Stmt::ExpectFx { remainder, .. }
+        | Stmt::Dbg { remainder, .. } => trmc_candidates_help(function_name, remainder, candidates),
+        Stmt::Join {
+            body, remainder, ..
+        } => {
+            let mut x = trmc_candidates_help(function_name, body, candidates.clone());
+            let y = trmc_candidates_help(function_name, remainder, candidates.clone());
+
+            x.extend(y);
+
+            x
+        }
+        Stmt::Ret(_) | Stmt::Jump(_, _) | Stmt::Crash(_, _) => candidates,
+    }
+}
+
+#[derive(Clone)]
+pub(crate) struct TrmcEnv<'a> {
+    hole_symbol: Symbol,
+    head_symbol: Symbol,
+    joinpoint_id: JoinPointId,
+    return_layout: InLayout<'a>,
+    ptr_return_layout: InLayout<'a>,
+
+    trmc_calls: MutMap<Symbol, Option<Call<'a>>>,
+}
+
+#[derive(Debug)]
+struct ConstructorInfo<'a> {
+    tag_layout: UnionLayout<'a>,
+    tag_id: TagIdIntType,
+    arguments: &'a [Symbol],
+}
+
+impl<'a> TrmcEnv<'a> {
+    #[inline(always)]
+    fn is_terminal_constructor(stmt: &Stmt<'a>) -> Option<ConstructorInfo<'a>> {
+        match stmt {
+            Stmt::Let(s1, expr, _layout, Stmt::Ret(s2)) if s1 == s2 => {
+                Self::get_contructor_info(expr)
+            }
+
+            _ => None,
+        }
+    }
+
+    fn get_contructor_info(expr: &Expr<'a>) -> Option<ConstructorInfo<'a>> {
+        if let Expr::Tag {
+            tag_layout,
+            tag_id,
+            arguments,
+        } = expr
+        {
+            let info = ConstructorInfo {
+                tag_layout: *tag_layout,
+                tag_id: *tag_id,
+                arguments,
+            };
+
+            Some(info)
+        } else {
+            None
+        }
+    }
+
+    fn is_recursive_expr(expr: &Expr<'a>, lambda_name: LambdaName<'_>) -> Option<Call<'a>> {
+        if let Expr::Call(call) = expr {
+            Self::is_recursive_call(call, lambda_name).then_some(call.clone())
+        } else {
+            None
+        }
+    }
+
+    fn is_recursive_call(call: &Call<'a>, lambda_name: LambdaName<'_>) -> bool {
+        match call.call_type {
+            CallType::ByName { name, .. } => {
+                // TODO are there other restrictions?
+                name == lambda_name
+            }
+            CallType::Foreign { .. } | CallType::LowLevel { .. } | CallType::HigherOrder(_) => {
+                false
+            }
+        }
+    }
+
+    fn ptr_write(
+        env: &mut Env<'a, '_>,
+        ptr: Symbol,
+        value: Symbol,
+        next: &'a Stmt<'a>,
+    ) -> Stmt<'a> {
+        let ptr_write = Call {
+            call_type: crate::ir::CallType::LowLevel {
+                op: LowLevel::PtrStore,
+                // update_mode: env.next_update_mode_id(),
+                update_mode: UpdateModeId::BACKEND_DUMMY,
```

Comment:
we always can. it just means optimizations are not applied here. but since we don't do any runtime checks for "can we perform an in-place mutation" here, morphic has no benefit anyway.